File: archive/s/swift/duplicate-character-counter.swift
import Foundation

func error() {
    print("Usage: please provide a string")
}

func duplicateCharacterCounter(string: [String]) {
    if string.isEmpty || string[0].isEmpty {
        error()
    } else {
        let input = string[0]
        var dict = [Character: Int]()

        for char in input{
        dict[char, default: 0] += 1
        }

        var flag = false
        for char in input{
            if let count = dict[char], count > 1 {
                flag = true
                print("\(char): \(count)")
                dict[char] = 0
            }
        }
        if !flag {
            print("No duplicate characters")
        }
    } 
}

let string = CommandLine.arguments.dropFirst().map { String($0) }
duplicateCharacterCounter(string: Array(string))

File: archive/f/f-sharp/DuplicateCharacterCounter.fs
open System

let usage = "Usage: please provide a string"

module DuplicateCharacterCounter =
    let private getDuplicateCounts =
        Seq.countBy id
        >> Seq.filter (fun (_, count) -> count > 1)
        >> Seq.map (fun (c, count) -> sprintf "%c: %d" c count)
        >> Seq.toList

    let private formatOutput =
        function
        | [] -> "No duplicate characters"
        | items -> String.concat "\n" items

    let run input =
        input |> getDuplicateCounts |> formatOutput |> Ok

module Helpers =
    let (|Empty|NonEmpty|) (s: string) =
        match s.Trim() with
        | "" -> Empty
        | trimmed -> NonEmpty trimmed

    let parseArgs argv =
        match argv with
        | [| Empty |] -> Error usage
        | [| NonEmpty input |] -> Ok input
        | _ -> Error usage

    let handleResults =
        function
        | Ok result ->
            printfn "%s" result
            0
        | Error msg ->
            eprintfn "%s" msg
            1

[<EntryPoint>]
let main argv =
    argv
    |> Helpers.parseArgs
    |> Result.bind DuplicateCharacterCounter.run
    |> Helpers.handleResults
